doom/.doom.d/config.org
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src
* Variable settings
#+begin_src emacs-lisp
(setq user-full-name "Sam Yousefzadegan Hedin"
      user-mail-address "sam.hedin@gmail.com")

(setq auth-sources '("/home/sam/.authinfo"))
#+end_src

Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:
+ `doom-font'
+ `doom-variable-pitch-font'
+ `doom-big-font' -- used for `doom-big-font-mode'; use this for
  presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string. You generally only need these two:
#+begin_src emacs-lisp
;; (setq doom-font (font-spec :family "Fira Code" :size 11.5 :weight 'normal)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 11.5))
(setq doom-font (font-spec :family "Fira Code" :size 19.5 :weight 'normal)
      doom-variable-pitch-font (font-spec :family "sans" :size 19.5))
#+end_src

#+begin_src emacs-lisp
(setq display-line-numbers-type nil)
(setq confirm-kill-emacs nil)
(setq which-key-idle-delay 0.001)

(defun sneaky-save-buffer (&rest _r)
  (unless (or  (string-match-p (regexp-quote "*") (buffer-name))
               (string-match-p (regexp-quote "pdf") (buffer-name)))
    (save-buffer)))

(dolist (f '(grep winum-select-window-by-number evil-switch-to-windows-last-buffer magit-status projectile-compile-project recompile))
  (advice-add f :before #'sneaky-save-buffer))
#+end_src
* Theme
There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:

#+begin_src emacs-lisp
(setq doom-theme 'doom-Iosvkem)
(let ((time  (string-to-number (format-time-string "%H"))))
  (if (or (< time 7) (> time 14))
      (load-theme 'doom-one t)
    (load-theme 'doom-one-light t)))

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")
#+end_src

* My private functions
#+begin_src emacs-lisp
(defun copy-current-file-path ()
  "Copy the current buffer file name to the clipboard."
  (interactive)
  (let ((filename (if (equal major-mode 'dired-mode)
                      default-directory
                    (buffer-file-name))))
    (when filename
      (kill-new filename)
      (message "Copied buffer file name '%s' to the clipboard." filename))))

(defun pdfgrep ()
  (interactive)
  (grep (format "pdfgrep --ignore-case --recursive --page-number '%s' ." (read-string "Enter term: "))))
#+end_src

#+begin_src emacs-lisp
;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(load! "keybinds.el")
(keybinds-mode)

(global-hl-line-mode 0)

(setq-default line-spacing 8)
(add-hook 'prog-mode-hook 'highlight-parentheses-mode)
(advice-add #'rainbow-delimiters-mode :override #'ignore)

(setq projectile-enable-caching nil)

(setenv "PATH" (concat (getenv "PATH") ":" "/home/sam/.local/bin/"))
(setq exec-path (append exec-path '("/home/sam/.local/bin")))
#+end_src

* Packages
** Treemacs
#+begin_src emacs-lisp
(after! treemacs
  (set-face-attribute 'treemacs-root-face nil :height 1.0  :underline nil)
  (setq treemacs-width 50)
  (treemacs-resize-icons 15))
#+end_src
** Centaur tabs
#+begin_src emacs-lisp
(after! centaur-tabs
  (setq centaur-tabs-set-close-button nil))
#+end_src
** Org-mode
#+begin_src emacs-lisp
(setq org-startup-with-latex-preview t)
(add-hook 'org-mode-hook 'LaTex-math-mode)
(after! org
  (setq-default fill-column 120)
  (auto-fill-mode)
  (set-face-attribute 'org-block-begin-line nil :inherit 'org-block :height 0.8 :background nil)
  (set-face-attribute 'org-block-end-line nil :inherit 'org-block :height 0.8 :background nil)
  (set-face-attribute 'org-meta-line nil :height 0.8 :background nil)
  (setq org-latex-toc-command "\\tableofcontents \\clearpage")

  (org-babel-do-load-languages
   'org-babel-load-languages
   '((python . t)
     (emacs-lisp .t)
     (julia . t)
     (latex . t)
     (rust . t)
     (sh . t))))

;; (setq org-latex-listings 'minted
;;       org-latex-packages-alist '(("" "minted"))
;;       org-latex-pdf-process
;;       '("pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
;;         "pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"))

(setq org-latex-minted-options '(("breaklines" "true")
                                 ("breakanywhere" "true")))

#+end_src
** Dash-docs
#+begin_src emacs-lisp
;; (dolist (f '("Julia" "Python_3" "NumPy" "SciPy"))
;;   (dash-docs-install-docset f))

;; (dolist (f '("scikit-learn" "PyTorch"))
;;   (dash-docs-install-user-docset f))

(setq dash-docs-docsets '("Julia" "Python 3" "NumPy" "SciPy" "scikit-learn" "PyTorch"))
(setq dash-docs-common-docsets '("Julia" "Python 3" "NumPy" "SciPy" "scikit-learn" "PyTorch"))
(setq large-file-warning-threshold 100000000)
#+end_src

** LSP-mode
#+begin_src emacs-lisp
(after! lsp
  (setq lsp-signature-render-documentation nil))
;; (setq lsp-signature-render-documentation nil)

;; (use-package lsp-haskell
;;   :config
;;   (setq lsp-haskell-process-path-hie "/home/sam/.local/bin/haskell-language-server-wrapper"))

(after! rustic
  (setq rustic-lsp-server 'rust-analyzer))
#+end_src

** Paren-face
#+begin_src emacs-lisp
(global-paren-face-mode)
(setq-default paren-face-regexp "[][(){}]")
(setq-default paren-face-modes (append '(rustic-mode org-mode python-mode) paren-face-modes))
#+end_src
** PDF-tools
#+begin_src emacs-lisp
;; Did pdf-tools break? Try
;; (pdf-tools-install)

(after! pdf-view
  (setq pdf-view-resize-factor 1.10)
  (setq pdf-view-midnight-colors '("#dddddd" . "#262829")))
#+end_src
** Dired
#+begin_src emacs-lisp
(add-hook 'dired-mode-hook
	  (lambda ()
	    (dired-hide-details-mode)
	    (dired-sort-toggle-or-edit)))
#+end_src

** APL
#+begin_src emacs-lisp
(defun em-gnu-apl-init ()
  (setq buffer-face-mode-face 'gnu-apl-default)
  (buffer-face-mode))

(add-hook 'gnu-apl-interactive-mode-hook 'em-gnu-apl-init)
(add-hook 'gnu-apl-mode-hook 'em-gnu-apl-init)
(defun gnu-apl-keyboard ()
  (interactive)
  (set-input-method "APL-Z"))
(add-hook 'gnu-apl-interactive-mode-hook 'gnu-apl-keyboard)
#+end_src
** Modeline, rich-minority and mini-modeline
#+begin_src emacs-lisp
(setq rm-blacklist "")
(rich-minority-mode)
(mini-modeline-mode)
#+end_src

** EAF
#+begin_src emacs-lisp
;; (use-package eaf
;;   :load-path "~/.emacs.d/site-lisp/emacs-application-framework" ; Set to "/usr/share/emacs/site-lisp/eaf" if installed from AUR
;;   :custom
;;   (eaf-browser-continue-where-left-off t)
;;   :config
;;   (eaf-setq eaf-browser-enable-adblocker "true")
;;   (eaf-bind-key scroll_up "C-n" eaf-pdf-viewer-keybinding)
;;   (eaf-bind-key scroll_down "C-p" eaf-pdf-viewer-keybinding)
;;   (eaf-bind-key take_photo "p" eaf-camera-keybinding)
;;   (eaf-bind-key nil "M-q" eaf-browser-keybinding))
#+end_src


emacs/.emacs.d/settings.org
* Visual stuff
#+BEGIN_SRC emacs-lisp
(set-face-attribute 'default nil :font "Source Code Pro" :height 115)
(blink-cursor-mode -1)
(scroll-bar-mode -1)
(tool-bar-mode -1)
(set-fringe-mode 5)
(menu-bar-mode -1)
(setq org-startup-indented t)
(setq org-startup-folded nil)
#+END_SRC


* Functions
#+begin_src emacs-lisp
(defun open-settings ()
  (interactive)
    (find-file "/home/sam/dotfiles/emacs/.emacs.d/settings.org"))
#+end_src

* Packages
** straight.el
#+BEGIN_SRC emacs-lisp
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
(straight-use-package 'use-package)
#+END_SRC

** Magit
#+begin_src emacs-lisp
(use-package magit
  :straight t
  :bind (("C-c s" . magit-status)))   
#+end_src
** Projectile
#+begin_src emacs-lisp
(use-package projectile
  :straight t
  :bind (("C-c p p" . projectile-switch-project)
         ("C-c p s" . projectile-grep)
         ("C-c p f" . projectile-find-file)))
  
(projectile-global-mode +1)
#+end_src
** avy
#+begin_src emacs-lisp
(use-package avy
  :straight t
  :bind (("C-." . avy-goto-char)))
#+end_src
** Selectrum
#+begin_src emacs-lisp
(use-package selectrum
:straight t)
(straight-use-package 'prescient)
(straight-use-package 'selectrum-prescient)
(selectrum-prescient-mode +1)
(prescient-persist-mode +1)
(selectrum-mode +1)
#+end_src
** Company
#+begin_src emacs-lisp
(use-package company
:straight t
:config
(setq company-minimum-prefix-length 1))
(add-hook 'after-init-hook 'global-company-mode)
#+end_src

** Smartparens
#+begin_src emacs-lisp
(use-package smartparens
:straight t)
(require 'smartparens)
(require 'smartparens-config)
(add-hook 'programming-mode-hook #'smartparens-mode)
(smartparens-global-mode t)
#+end_src
** Which-key
#+begin_src emacs-lisp
(use-package which-key
:straight t
:config
(setq which-key-idle-delay 0.001))
(which-key-mode)
#+end_src
** Crux
#+begin_src emacs-lisp
(use-package crux
:straight t
:bind (("C-c f" . crux-recentf-find-file)
       ("C-o" . crux-smart-open-line)
      ("C-S-o" . crux-smart-open-line-above)))
#+end_src
** Dumb-jump
#+begin_src emacs-lisp
(use-package dumb-jump
:straight t)
(add-hook 'xref-backend-functions #'dumb-jump-xref-activate)
#+end_src
** Winum
#+begin_src emacs-lisp
(use-package winum
:straight t
:bind (
("C-2" . winum-select-window-2)
("C-3" . winum-select-window-3)
("C-1" . winum-select-window-1)))
(require 'winum)
(winum-mode)
#+end_src
* Binds
#+begin_src emacs-lisp

#+end_src


doom/.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq user-full-name "Sam Yousefzadegan Hedin"
      user-mail-address "sam.hedin@gmail.com")

(setq auth-sources '("/home/sam/.authinfo"))

;; (setq doom-font (font-spec :family "Fira Code" :size 11.5 :weight 'normal)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 11.5))
(setq doom-font (font-spec :family "Fira Code" :size 19.5 :weight 'normal)
      doom-variable-pitch-font (font-spec :family "sans" :size 19.5))

(setq display-line-numbers-type nil)
(setq confirm-kill-emacs nil)
(setq which-key-idle-delay 0.001)

(defun sneaky-save-buffer (&rest _r)
  (unless (or  (string-match-p (regexp-quote "*") (buffer-name))
               (string-match-p (regexp-quote "pdf") (buffer-name)))
    (save-buffer)))

(dolist (f '(grep winum-select-window-by-number evil-switch-to-windows-last-buffer magit-status projectile-compile-project recompile))
  (advice-add f :before #'sneaky-save-buffer))

(setq doom-theme 'doom-Iosvkem)
(let ((time  (string-to-number (format-time-string "%H"))))
  (if (or (< time 7) (> time 14))
      (load-theme 'doom-one t)
    (load-theme 'doom-one-light t)))

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

(defun copy-current-file-path ()
  "Copy the current buffer file name to the clipboard."
  (interactive)
  (let ((filename (if (equal major-mode 'dired-mode)
                      default-directory
                    (buffer-file-name))))
    (when filename
      (kill-new filename)
      (message "Copied buffer file name '%s' to the clipboard." filename))))

(defun pdfgrep ()
  (interactive)
  (grep (format "pdfgrep --ignore-case --recursive --page-number '%s' ." (read-string "Enter term: "))))

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(load! "keybinds.el")
(keybinds-mode)

(global-hl-line-mode 0)

(setq-default line-spacing 8)
(add-hook 'prog-mode-hook 'highlight-parentheses-mode)
(advice-add #'rainbow-delimiters-mode :override #'ignore)

(setq projectile-enable-caching nil)

(after! treemacs
  (set-face-attribute 'treemacs-root-face nil :height 1.0  :underline nil)
  (setq treemacs-width 50)
  (treemacs-resize-icons 15))

(after! centaur-tabs
  (setq centaur-tabs-set-close-button nil))

(setq org-startup-with-latex-preview t)
;; (add-hook 'org-mode-hook 'LaTex-math-mode)

(after! org
  (setq-default fill-column 120)
  (auto-fill-mode)
  (set-face-attribute 'org-block-begin-line nil :inherit 'org-block :height 0.8 :background nil)
  (set-face-attribute 'org-block-end-line nil :inherit 'org-block :height 0.8 :background nil)
  (set-face-attribute 'org-meta-line nil :height 0.8 :background nil)
  (setq org-latex-toc-command "\\tableofcontents \\clearpage")

  (org-babel-do-load-languages
   'org-babel-load-languages
   '((python . t)
     (emacs-lisp .t)
     (julia . t)
     (latex . t)
     (jupyter . t)
     (rust . t)
     (sh . t))))
(setq org-babel-default-header-args:jupyter-julia '((:async . "yes")
						    (:exports . "both")
						    (:results . "scalar")))
;; (setq org-latex-listings 'minted
;;       org-latex-packages-alist '(("" "minted"))
;;       org-latex-pdf-process
;;       '("pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
;;         "pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"))

(setq org-latex-minted-options '(("breaklines" "true")
                                 ("breakanywhere" "true")))

;;  Install dash docsets with these functions.
;; (dolist (f '("Julia" "Python_3" "NumPy" "SciPy" "Mono"))
;;   (dash-docs-install-docset f))

;; (dolist (f '("scikit-learn" "PyTorch"))
;;   (dash-docs-install-user-docset f))

(add-hook 'csharp-mode-hook
          (lambda ()
            (setq-local dash-docs-docsets '("Mono"))))

(add-hook 'python-mode-hook
          (lambda ()
            (setq-local dash-docs-docsets '("Python 3" "NumPy" "SciPy" "scikit-learn"))))

(add-hook 'julia-mode-hook
          (lambda ()
            (setq-local dash-docs-docsets '("Julia"))))
(add-hook 'org-mode-hook
          (lambda ()
            (setq-local dash-docs-docsets '("Julia"))))

(after! lsp
  (setq lsp-signature-render-documentation nil))

;; (add-hook 'csharp-mode-hook
;;           (lambda ()))
(setq lsp-csharp-server-path "/run/current-system/sw/bin/omnisharp")

(after! rustic
  (setq rustic-lsp-server 'rust-analyzer))
;; (setq rustic-compile-display-method 'display-buffer-other-frame)

(setq lsp-julia-default-environment "~/.julia/environments/v1.5")

(global-paren-face-mode)
(setq-default paren-face-regexp "[][(){}]")
(setq-default paren-face-modes (append '(rustic-mode org-mode python-mode) paren-face-modes))

;; Did pdf-tools break? Try
;; (pdf-tools-install)

(after! pdf-view
  (setq pdf-view-resize-factor 1.10)
  (setq pdf-view-midnight-colors '("#dddddd" . "#262829")))

(add-hook 'dired-mode-hook
	  (lambda ()
	    (dired-hide-details-mode)
	    (dired-sort-toggle-or-edit)))

(setq rm-blacklist "")
(rich-minority-mode)
(mini-modeline-mode)

(defun remove-dos-eol ()
  "Do not show ^M in files containing mixed UNIX and DOS line endings."
  (interactive)
  (setq buffer-display-table (make-display-table))
  (aset buffer-display-table ?\^M []))

(add-hook 'csharp-mode-hook 'remove-dos-eol)
